storage/response_cache.go

```diff
@@ -158,32 +164,43 @@ func (rc *ResponseCache) CopyFrom(e engine.Engine, originRaftID proto.RaftID) er
 	})
 }
 
-// PutResponse writes a response to the cache for the specified cmdID.
-func (rc *ResponseCache) PutResponse(e engine.Engine, cmdID proto.ClientCmdID, reply proto.Response) error {
+// PutResponse writes a response and an error associated with it to the
+// cache for the specified cmdID.
+func (rc *ResponseCache) PutResponse(e engine.Engine, cmdID proto.ClientCmdID, reply proto.Response, replyErr error) error {
 	// Do nothing if command ID is empty.
 	if cmdID.IsEmpty() {
 		return nil
 	}
+
 	// Write the response value to the engine.
-	var err error
-	if rc.shouldCacheResponse(reply) {
+	if rc.shouldCacheResponse(reply, replyErr) {
+		// Write the error into the reply before caching.
+		header := reply.Header()
+		header.SetGoError(replyErr)
+		// Be sure to clear it when you're done!
+		defer func() { header.Error = nil }()
```

Comment:
It wouldn't help here though, since we're setting this error for the purpose of serialization.